### PR TITLE
[JENKINS-64423, JENKINS-46618] Resolve all core caused startup Illegal reflective access warnings

### PR DESCRIPTION
**Pull request by timja**
See [JENKINS-64423](https://issues.jenkins-ci.org/browse/JENKINS-64423).
See [JENKINS-46618](https://issues.jenkins-ci.org/browse/JENKINS-46618).

https://groups.google.com/g/jenkinsci-dev/c/nEw4TLDcZLw/m/_7LVO0GnAAAJ

There's still illegal reflective access warnings on startup that require library upgrades (or xstream might need core / plugin tweaks):
* Guice to 5.0 (only in beta atm)
* Groovy to 3.x
* some xstream ones only triggered in certain circumstances, haven't managed to trigger it in a debugger atm 😢 

<details>

<summary>xstream</summary>

```
WARNING: Illegal reflective access by com.thoughtworks.xstream.converters.reflection.FieldDictionary (file:/Users/timja/.jenkins/war/WEB-INF/lib/xstream-1.4.15.jar) to field java.util.AbstractCollection.MAX_ARRAY_SIZE
WARNING: Illegal reflective access by com.thoughtworks.xstream.converters.reflection.FieldDictionary (file:/Users/timja/.jenkins/war/WEB-INF/lib/xstream-1.4.15.jar) to field java.util.AbstractList.modCount
WARNING: Illegal reflective access by com.thoughtworks.xstream.converters.reflection.FieldDictionary (file:/Users/timja/.jenkins/war/WEB-INF/lib/xstream-1.4.15.jar) to field java.util.Collections$EmptyList.serialVersionUID
WARNING: Illegal reflective access by com.thoughtworks.xstream.core.util.SerializationMembers (file:/Users/timja/.jenkins/war/WEB-INF/lib/xstream-1.4.15.jar) to method java.util.Collections$EmptyList.readResolve()
WARNING: Illegal reflective access by com.thoughtworks.xstream.converters.reflection.FieldDictionary (file:/Users/timja/.jenkins/war/WEB-INF/lib/xstream-1.4.15.jar) to field java.util.AbstractCollection.MAX_ARRAY_SIZE
WARNING: Illegal reflective access by com.thoughtworks.xstream.converters.reflection.FieldDictionary (file:/Users/timja/.jenkins/war/WEB-INF/lib/xstream-1.4.15.jar) to field java.util.AbstractList.modCount
WARNING: Illegal reflective access by com.thoughtworks.xstream.converters.reflection.FieldDictionary (file:/Users/timja/.jenkins/war/WEB-INF/lib/xstream-1.4.15.jar) to field java.util.AbstractCollection.MAX_ARRAY_SIZE
WARNING: Illegal reflective access by com.thoughtworks.xstream.converters.reflection.FieldDictionary (file:/Users/timja/.jenkins/war/WEB-INF/lib/xstream-1.4.15.jar) to field java.util.AbstractList.modCount
WARNING: Illegal reflective access by com.thoughtworks.xstream.converters.reflection.FieldDictionary (file:/Users/timja/.jenkins/war/WEB-INF/lib/xstream-1.4.15.jar) to field java.util.Collections$EmptyList.serialVersionUID
WARNING: Illegal reflective access by com.thoughtworks.xstream.core.util.SerializationMembers (file:/Users/timja/.jenkins/war/WEB-INF/lib/xstream-1.4.15.jar) to method java.util.Collections$EmptyList.readResolve()
```

</details>

### Proposed changelog entries

* Entry 1: Issue, Human-readable Text
* ...

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/ClassicPluginStrategy.java
Comment (timja) [inline core/src/main/java/hudson/ClassicPluginStrategy.java:636]: wondering if this `ClassLoaderReflectionToolkit` could be removed, assuming everything goes through `AntWithFindResourceClassLoader` then this line never gets called.

(from my minimal plugin set in my test Jenkins this assumption is always true)

any thoughts? this comment applies throughout the PR

File: core/src/main/java/hudson/ExtensionFinder.java
Comment (timja) [inline core/src/main/java/hudson/ExtensionFinder.java:493]: this call seems to just check if the class is null?

image

If I add a null check IDE tell's me that it's impossible for it to be null at this point.

Removing this check doesn't break any tests

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (timja) [inline core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:17]: changes in this class are mostly lazy initialisation, to stop warnings if the code isn't used, possible this class could be deprecated as I can't see anything that requires it, but I've left fallbacks in that call this code for extra safety

File: core/src/main/java/hudson/ClassicPluginStrategy.java
Comment (jglick) [reply core/src/main/java/hudson/ClassicPluginStrategy.java:636]: Not sure. May apply in `JenkinsRule` under exotic conditions?

File: core/src/main/java/hudson/PluginManager.java
Comment (jglick) [inline core/src/main/java/hudson/PluginManager.java:2159]: To avoid duplication, maybe revert all these changes at call sites, and encode the desired nonreflective behavior into `ClassLoaderReflectionToolkit`? (For clarity, could introduce an `interface` implemented by `AntWithFindResourceClassLoader` which defines the required methods.)

File: core/src/main/java/hudson/ExtensionFinder.java
Comment (jglick) [reply core/src/main/java/hudson/ExtensionFinder.java:493]: Not sure what version of Java you are looking at. In Java 8

```java
protected final void resolveClass(Class<?> c) {
    resolveClass0(c);
}

private native void resolveClass0(Class<?> c);
```

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (jglick) [inline core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:53]: Not thread-safe. You could either make this and similar methods `synchonized`, or use a `private static` nested holder class with `final` variables to implement lazy initialization (which would make for a simpler patch overall I guess).

**Review by jglick — COMMENTED** [review]
Looks right in principle.

File: core/src/main/java/hudson/ClassicPluginStrategy.java
Comment (timja) [reply core/src/main/java/hudson/ClassicPluginStrategy.java:636]: it does seem to apply in Jenkinsfile :(, see my happiness in my commit message here:
https://github.com/jenkinsci/jenkins/pull/5110/commits/555e516ea72e3319ea06caca0facb7740c1caf48

File: core/src/main/java/hudson/ExtensionFinder.java
Comment (timja) [reply core/src/main/java/hudson/ExtensionFinder.java:493]: Java 11 Azul

File: core/src/main/java/hudson/ClassicPluginStrategy.java
Comment (jglick) [reply core/src/main/java/hudson/ClassicPluginStrategy.java:636]: :man_shrugging: if you can verify that only `AntWithFindResourceClassLoader` is ever used at this point, including in various test harnesses, we could probably simplify a bunch of code, and put in some `assert`ions.

File: core/src/main/java/hudson/PluginManager.java
Comment (timja) [reply core/src/main/java/hudson/PluginManager.java:2159]: yup was thinking about that, the only reason I didn't do that was I was looking to see if we could retire it completely

File: core/src/main/java/hudson/ExtensionFinder.java
Comment (jglick) [reply core/src/main/java/hudson/ExtensionFinder.java:493]: Maybe this became obsolete somehow in 11, but still appears to do something real in 8, so suggest reverting.

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (timja) [reply core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:53]: second method sounds good :+1:

thanks I was worried about that bit

File: core/src/main/java/hudson/ClassicPluginStrategy.java
Comment (timja) [reply core/src/main/java/hudson/ClassicPluginStrategy.java:636]: Well I can verify at runtime it's fine, but in tests (at least AssetManagerTest it's not fine, it runs with a classloader built into the JDK, AppClassLoader I think.

Comment (jglick) [reply core/src/main/java/hudson/ClassicPluginStrategy.java:636]: So sounds like we still need `ClassLoaderReflectionToolkit` at least for now.

File: core/src/main/java/hudson/ExtensionFinder.java
Comment (timja) [reply core/src/main/java/hudson/ExtensionFinder.java:493]: The code that called this is removed in java 9, https://bugs.openjdk.java.net/browse/JDK-8057777

I'm not sure if it was doing something in 8 or not, the code was removed because it wasn't doing anything but it doesn't say why, it couldn't been something in 9 that meant it wasn't used any more.

Comment (res0nance) [reply core/src/main/java/hudson/ExtensionFinder.java:493]: I can't seem to see where the definition of resolveClass0 lies to confirm no side-effects. It seems difficult to confirm that removing this is safe.

> it couldn't been something in 9 that meant it wasn't used any more.

How did you arrive at this conclusion?

Even if that is the case there is a possibility it was removed in a jdk8 update but which update exactly, there might be breakages on systems with older versions of jdk8 possibly.

Comment (timja) [reply core/src/main/java/hudson/ExtensionFinder.java:493]: I read the code, but the openjdk repo is only tagged up to v120 or so of java 8.

the code definitely does nothing on 9.

If people think it’s still worth keeping I can re-add it but only for java 8

File: core/src/main/java/hudson/ClassicPluginStrategy.java
Comment (jglick) [inline core/src/main/java/hudson/ClassicPluginStrategy.java:635]: Prefer to revert unnecessary hunks to keep the diff minimal.

File: core/src/main/java/hudson/ClassicPluginStrategy.java
Comment (jglick) [inline core/src/main/java/hudson/ClassicPluginStrategy.java:688]:
```suggestion
                    URL url = ClassLoaderReflectionToolkit._findResource(pw.classLoader, name);
```

and suggest reverting entire hunk anyway

File: core/src/main/java/hudson/ExtensionFinder.java
Comment (jglick) [reply core/src/main/java/hudson/ExtensionFinder.java:493]: In http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/ `jdk/src/share/native/java/lang/ClassLoader.c` as of `jdk8u282-b07` has

```cpp
JNIEXPORT void JNICALL
Java_java_lang_ClassLoader_resolveClass0(JNIEnv *env, jobject this,
                                         jclass cls)
{
    if (cls == NULL) {
        JNU_ThrowNullPointerException(env, 0);
        return;
    }

    JVM_ResolveClass(env, cls);
}
```

which is defined in `hotspot/src/share/vm/prims/jvm.cpp` as

```cpp
JVM_ENTRY(void, JVM_ResolveClass(JNIEnv* env, jclass cls))
  JVMWrapper("JVM_ResolveClass");
  if (PrintJVMWarnings) warning("JVM_ResolveClass not implemented");
JVM_END
```

and this code does not appear to have been edited since at least 2007! I suppose the Javadoc remains because another JVM could in principle implement this, but seeing as this was just some old heuristic to work around alleged Guice problems, I concur that this is safe to remove.

File: core/src/main/java/hudson/PluginManager.java
Comment (jglick) [inline core/src/main/java/hudson/PluginManager.java:2194]: revert hunks

File: core/src/main/java/hudson/PluginManager.java
Comment (jglick) [inline core/src/main/java/hudson/PluginManager.java:2213]: could revert

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (jglick) [inline core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:16]:
```suggestion
 * <p>
```

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (jglick) [inline core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:20]:
```suggestion
 * <p>
```

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (jglick) [inline core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:22]:
```suggestion
 * <p>
```

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (jglick) [inline core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:18]: Is this supposed to be a class name? If so, use `@link` or (if 9+) `@code`. Or some other code name? `@code`. Otherwise, avoid CamelCase.

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (jglick) [inline core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:23]:
```suggestion
 * Note: Currently there is no known production use-case for the fallback case of accessing these methods via reflection:
```

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (jglick) [inline core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:24]:
```suggestion
 * the {@code JenkinsRule} tests use a different classloader, but once that is made consistent with production Jenkins we can
```

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (jglick) [inline core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:25]: (also prefer one sentence or major phrase per line, as in AsciiDoc)

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (jglick) [inline core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:205]: Wonder why these are not using `invoke`.

File: core/src/main/java/jenkins/model/AssetManager.java
Comment (jglick) [inline core/src/main/java/jenkins/model/AssetManager.java:94]: revert

File: core/src/main/java/jenkins/util/AntWithFindResourceClassLoader.java
Comment (jglick) [inline core/src/main/java/jenkins/util/AntWithFindResourceClassLoader.java:50]:
```suggestion
        return findLoadedClass(name);
```

**Review by jglick — APPROVED** [review]

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (timja) [reply core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:18]:
```suggestion
 * to prevent illegal reflective access errors on Java 11+
```

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (timja) [reply core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:205]: hmm isn't it using invoke? or do you mean something else?

File: core/src/main/java/jenkins/model/AssetManager.java
Comment (timja) [reply core/src/main/java/jenkins/model/AssetManager.java:94]: why? the try catch was removed so indentation changed

File: core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java
Comment (jglick) [reply core/src/main/java/jenkins/ClassLoaderReflectionToolkit.java:205]: I meant `ClassLoaderReflectionToolkit.invoke` rather than `Method.invoke`.

**Review by res0nance — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Review by oleg-nenashev — APPROVED** [review]
Looks good. `ClassLoaderReflectionToolkit`  is not used outside of the Jenkins core, so no need to experiment with custom classloaders defined in plugins.

**Pull request merged**
